llvm/lib/Transforms/Scalar/SROA.cpp

```diff
@@ -2853,6 +2853,11 @@ class AllocaSliceRewriter : public InstVisitor<AllocaSliceRewriter, bool> {
 
   bool visitLoadInst(LoadInst &LI) {
     LLVM_DEBUG(dbgs() << "    original: " << LI << "\n");
+
+    // load atomic vector would be generated, which is illegal
+    if (LI.isAtomic() && NewAI.getAllocatedType()->isVectorTy())
```

Comment:
Is it illegal from LangRef's perspective?

Reply:
It is at least illegal from the perspective of IR/Verifier.

Reply:
It looks like this is written [here](https://llvm.org/docs/LangRef.html#id211:~:text=results%20when%20they%20may%20see%20multiple%20atomic%20stores.-,The%20type%20of%20the%20pointee%20must%20be%20an%20integer%2C%20pointer%2C%20or%20floating%2Dpoint%20type%20whose%20bit%20width%20is%20a%20power%20of%20two%20greater%20than%20or%20equal%20to%20eight%20and%20less%20than%20or%20equal%20to%20a%20target%2Dspecific%20size%20limit.,-align%20must%20be%20explicitly%20specified%20on%20atomic%20loads).

Reply:
The restriction is dumb and we should relax it. Instead of just hardcoding isVectorTy here, should have some kind of LoadInst::isValidAtomicType or something

Reply:
Something like `bool LoadInst::isValidAtomicTy(AllocaInst &AI) const`? AllocaSliceRewriter needs to look at the associated `alloca`.

Reply:
No alloca reference, just the type. The implementation should match Verifier::visitLoadInst, the isIntOrIntPtrTy || isFloatingPointTy and checkAtomicMemAccessSize (also add some non-atomic size tests?)

Reply:
No, you need to ensure the atomic is a valid type for an atomic load. 

Alternatively you can do the load with the equivalent sized type and then bitcast (which is why this restriction is dumb in the first place, the lowering can always do the same)

Reply:
We are ensuring it is a valid type in the case under question, by checking the alloca's type.

Reply:
We don't want any loads that are atomic to have a vector type, regardless of whether the atomic itself is already valid.